### PR TITLE
Enable option for subword regularization in `XLMRobertaTokenizer`

### PR DESCRIPTION
**Pull request by PhilipMay**
# What does this PR do?
I would like to use [subword regularization](https://github.com/google/sentencepiece#subword-regularization-and-bpe-dropout) from [google/sentencepiece](https://github.com/google/sentencepiece). The reason is that it might be used to improve downstream task performance.

Since `XLMRobertaTokenizer` already uses `SentencePieceProcessor` from `google/sentencepiece` there are only some minor modifications needed.

3 additional parameters are added to the constructor of `XLMRobertaTokenizer`. This are:
```python
        enable_sampling=False,
        nbest_size=-1,
        alpha=0.1,
```
The default values are selected so that this is no breaking change.

In the `_tokenize(self, text)` function there was a call to `self.sp_model.EncodeAsPieces(text)`. This call does ignore the parameters for subword regularization. That is why it had to be replaced by a call to `self.sp_model.encode(text, out_type=str)`.

Since `XLMRobertaTokenizerFast` is an independent implentation which does not use `google/sentencepiece` it is not in the scope of the PR to add subword regularization to the fast tokenizer.

## To-do
- [x] check if tests pass
- [x] check if tests can / should be added
- [x] add a link to a page where we can see all kwargs

## Who can review?
@LysandreJik @stefan-it

### CODE REVIEW

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 1]
```diff
@@ -94,6 +94,18 @@ class XLMRobertaTokenizer(PreTrainedTokenizer):
...
+            Enable subword regularization.
+        nbest_size (:obj:`int`, `optional`, defaults to -1):
+            Sampling parameters for unigram. Invalid for BPE-Dropout.
+            nbest_size = {0,1}: No sampling is performed.
+            nbest_size > 1: samples from the nbest_size results.
+            nbest_size < 0: assuming that nbest_size is infinite and samples
+            from the all hypothesis (lattice) using
+            forward-filtering-and-backward-sampling algorithm.
+        alpha (:obj:`float`, `optional`, defaults to 0.1):
+            Soothing parameter for unigram sampling, and dropout probability of
```
Comment (stefan-it) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:107]:
```suggestion
            Smoothing parameter for unigram sampling, and dropout probability of
```

**Comment by PhilipMay** [comment]
I added a test, everything is green and IMO ready for review. @LysandreJik @stefan-it

**Review by LysandreJik — COMMENTED** [review]
Cool, this is a nice addition! Thanks for adding tests to go with it.

Do you think it would be possible to add this to the fast tokenizer as well? cc @n1t0

**Comment by PhilipMay** [comment]
@LysandreJik and @n1t0 I think that would be a good idea but IMO it should not be done in the scope of this PR.

Because the slow tokenizer just delegates the work to [google/sentencepiece](https://github.com/google/sentencepiece) this PR is very easy but adding that to the Rust Tokenizer would be way more work afaik.

**Comment by PhilipMay** [comment]
Hey @LysandreJik and @n1t0 
I think this PR is somehow stuck...
AFAIK my change is ok for you. 
What about merging it and moving the part for the fast tokenizer to a seperate issue?

**Review by LysandreJik — APPROVED** [review]
Yes, LGTM! Thanks again for adding the tests. We should think about adding this to the fast tokenizers eventually, as discussed with @n1t0.

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 2]
```diff
@@ -252,7 +270,7 @@ def get_vocab(self):
         return vocab
 
     def _tokenize(self, text):
-        return self.sp_model.EncodeAsPieces(text)
+        return self.sp_model.encode(text, out_type=str)
```
Comment (LysandreJik) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:273]: Is this change necessary, and does it change anything of the current behavior?

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 3]
```diff
@@ -128,10 +139,17 @@ def __init__(
             cls_token=cls_token,
             pad_token=pad_token,
             mask_token=mask_token,
+            enable_sampling=enable_sampling,
+            nbest_size=nbest_size,
+            alpha=alpha,
```
Comment (LysandreJik) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:144]: Talking about it with @sgugger, we think this might be better suited in some `**sp_model_kwargs` or something along those lines, that we could then add to all SentencePiece-based tokenizers. How does that sound?

**Review by LysandreJik — COMMENTED** [review]
We have a few comments before merging

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py (hunk 2 above)
Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:273]: Yes - see my initial posting here:

> In the `_tokenize(self, text)` function there was a call to `self.sp_model.EncodeAsPieces(text)`. This call does ignore the parameters for subword regularization. That is why it had to be replaced by a call to `self.sp_model.encode(text, out_type=str)`.

Without this change the parameters set in the constructor (`enable_sampling`, `nbest_size` and `alpha`) would be ignored and no augmentation would be done.

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py (hunk 3 above)
Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py]: Ok - cool. I made the change.

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py (hunk 2 above)
Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:270]: IMO this is resolved.

**Comment by PhilipMay** [comment]
@LysandreJik 
- Requested changes are made and marked as resolved. 
- Inline questions are answered and marked as resolved.
- CI is green.

IMO ready for merge.

**Review by stefan-it — APPROVED** [review]

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 4]
```diff
@@ -94,6 +94,17 @@ class XLMRobertaTokenizer(PreTrainedTokenizer):
             modeling. This is the token which the model will try to predict.
         additional_special_tokens (:obj:`List[str]`, `optional`, defaults to :obj:`["<s>NOTUSED", "</s>NOTUSED"]`):
             Additional special tokens used by the tokenizer.
+        sp_model_kwargs (:obj:`dict`, `optional`, defaults to :obj:`None`):
+            Will be passed to the SentencePieceProcessor ``__init__()`` method. Can be used, among other things, to
```
Comment (sgugger) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:98]:
```suggestion
            Will be passed to the ``.SentencePieceProcessor.__init__()`` method. Can be used, among other things, to
```
Also, adding a link to a page where we can see all kwargs would be awesome!

**Review by sgugger — APPROVED** [review]
Thanks for adapting your PR, this looks good to me apart from the docstring, which we can make a little bit better :-)

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py (hunk 4 above)
Comment (stefan-it) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py]: Documentation of Python API is really bad, I always have to look at the spm code directly:

https://github.com/google/sentencepiece/blob/351600c2971401f4e849147579aa1b5d42f614e1/python/src/sentencepiece/sentencepiece.i#L224

It shows some possible args :)

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 5]
```diff
@@ -94,6 +94,17 @@ class XLMRobertaTokenizer(PreTrainedTokenizer):
...
+            set:
+
+            - ``enable_sampling``: Enable subword regularization.
+            - ``nbest_size``: Sampling parameters for unigram. Invalid for BPE-Dropout. nbest_size = {0,1}: No sampling
+              is performed. nbest_size > 1: samples from the nbest_size results. nbest_size < 0: assuming that
+              nbest_size is infinite and samples from the all hypothesis (lattice) using
+              forward-filtering-and-backward-sampling algorithm.
```
[the remark this answers is not in the document]
Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py]: I reformatted it and added a link to the python API of SentencePiece.

Marking as resolved now.

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py (hunk 4 above)
Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py]: I added a link to the Python API root folder but not to the Python script or code line - since it can change from version to version.

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 6]
```diff
@@ -94,6 +94,19 @@ class XLMRobertaTokenizer(PreTrainedTokenizer):
             modeling. This is the token which the model will try to predict.
         additional_special_tokens (:obj:`List[str]`, `optional`, defaults to :obj:`["<s>NOTUSED", "</s>NOTUSED"]`):
             Additional special tokens used by the tokenizer.
+        sp_model_kwargs (:obj:`dict`, `optional`, defaults to :obj:`None`):
+            Will be passed to the ``SentencePieceProcessor.__init__()`` method. The `Python wrapper for SentencePiece
+            <https://github.com/google/sentencepiece/tree/master/python>`_ can be used, among other things, to set:
```
Comment (sgugger) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:99]:
```suggestion
            <https://github.com/google/sentencepiece/tree/master/python>`__ can be used, among other things, to set:
```

File: src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py [hunk 7]
```diff
@@ -94,6 +94,19 @@ class XLMRobertaTokenizer(PreTrainedTokenizer):
...
+
+            - ``enable_sampling``: Enable subword regularization.
+            - ``nbest_size``: Sampling parameters for unigram. Invalid for BPE-Dropout.
+
+              - ``nbest_size = {0,1}``: No sampling is performed.
+              - ``nbest_size > 1``: samples from the nbest_size results.
+              - ``nbest_size < 0``: assuming that nbest_size is infinite and samples from the all hypothesis (lattice)
+                using forward-filtering-and-backward-sampling algorithm.
```
Comment (sgugger) [inline src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:107]: You should intent at 4 with a new line before and after otherwise sphinx might not be very happy with you.

Comment (PhilipMay) [reply src/transformers/models/xlm_roberta/tokenization_xlm_roberta.py:107]: Fixed.

**Comment by PhilipMay** [comment]
@sgugger all green again :-)

**Review by sgugger — APPROVED** [review]
Perfect! If you feel up to the task, I think all (slow) sentencepiece-based tokenizers could benefit from this addition.

**Comment by PhilipMay** [comment]
> Perfect! If you feel up to the task, I think all (slow) sentencepiece-based tokenizers could benefit from this addition.

see #11417
